### PR TITLE
Allow ThreadPoolExecutor for use in AWS Lambda functions

### PR DESCRIPTION
**Pull request by doc-E-brown**
This PR addresses #339 in providing a flag within the generate_clk function signature to use a ThreadPoolExecutor as opposed to ProcessPoolExecutor.  This change is propagated throughout the function signatures that implement generate_clk as well as the cli and benchmark cli. 

Documentation has also been updated to accommodate the change.

### CODE REVIEW

**Comment by doc-E-brown** [comment]
Build failures are mostly due to a missing codecov token in the build pipeline as well as one pytest timeout.

File: clkhash/cli.py [hunk 1]
```diff
@@ -141,8 +141,9 @@ def cli(verbose):
 @click.option('--no-header', default=False, is_flag=True, help="Don't skip the first row")
 @click.option('--check-header', default=True, type=bool, help="If true, check the header against the schema")
 @click.option('--validate', default=True, type=bool, help="If true, validate the entries against the schema")
+@click.option('--multiprocessing', default=True, type=bool, help="If true, use multiprocessing to generate the keys from the secret")
```
Comment (hardbyte) [inline clkhash/cli.py]:
```suggestion
@click.option('--multiprocessing', default=True, type=bool, help="If true, use multiprocessing to generate encodings")
```

File: clkhash/clk.py [hunk 2]
```diff
@@ -75,6 +76,11 @@ def generate_clk_from_csv(input_f,  # type: TextIO
             header but it should not be checked against the schema.
         :param bool progress_bar: Set to `False` to disable the progress
             bar.
+        :param bool use_multiprocessing: Set to `False` to use threading to
+            generate keys for each identifier from the secret.  By default
+            multiprocessing is used to generate the keys.  Set this flag to
+            `False` if executing on a system that is not capable of spawning
+            subprocesses such as AWS Lambda functions.
```
Comment (hardbyte) [inline clkhash/clk.py]:
```suggestion
        :param bool use_multiprocessing: Set to `False` to use threading instead
            of multiprocessing to generate encodings for each record. Set this flag 
            to `False` if executing on a system that is not capable of spawning
            subprocesses such as AWS Lambda functions.
```

File: clkhash/clk.py [hunk 3]
```diff
@@ -154,7 +165,10 @@ def generate_clks(pii_data,  # type: Sequence[Sequence[str]]
     futures = []
 
     # Compute Bloom filter from the chunks and then serialise it
-    with concurrent.futures.ProcessPoolExecutor() as executor:
+    pool_executor = ProcessPoolExecutor if use_multiprocessing else \
+        ThreadPoolExecutor # type: Union[Type[ProcessPoolExecutor], Type[ThreadPoolExecutor]]
+
+    with pool_executor() as executor:
```
Comment (hardbyte) [inline clkhash/clk.py]: Nice

File: docs/cli.rst [hunk 4]
```diff
@@ -187,12 +187,7 @@ A corresponding hashing schema can be generated as well::
             "weight": 0
           }
         },
-        {
-          "identifier": "NAME freetext",
-          "format": {
-            "type": "string",
-            "encoding": "utf-8",
-            "case": "mixed",
+        {s
```
Comment (hardbyte) [inline docs/cli.rst]: Assume this was a mistake?

File: docs/cli.rst [hunk 5]
```diff
@@ -250,6 +243,16 @@ As a rule of thumb a single modern core will hash around 1M entities in about 20
...
 as a basic sanity check - ensure the CLK's popcount is not around 0 or 1024.
 
+By default the benchmark uses multiprocessing to generate the clks, the performance difference by switching
+to multi-threading mode can also be determined by passing the command line option ``--multiprocessing 0``.
+
+    python -m clkhash.cli benchmark --multiprocessing 0
+    generating CLKs: 100%                 10.0K/10.0K [00:05<00:00, 1.90Kclk/s, mean=406, std=20.4]
+     10000 hashes in 5.285763 seconds. 1.89 KH/s
+
+The performance reduction in using multithreading mode is significant however it does allo clks to be
```
Comment (hardbyte) [inline docs/cli.rst]:
```suggestion
The performance reduction in using multithreading mode is significant however it does allow clks to be 
```

File: docs/cli.rst (hunk 4 above)
Comment (doc-E-brown) [reply docs/cli.rst]: Yes, not sure how this ended up here.  I will remove

File: clkhash/clk.py (hunk 3 above)
Comment (doc-E-brown) [reply clkhash/clk.py]: Thanks!

File: docs/cli.rst (hunk 4 above)
Comment (hardbyte) [reply docs/cli.rst]: The deletion as well thanks

Comment (doc-E-brown) [reply docs/cli.rst]: Sorry! Corrected

**Comment by hardbyte** [comment]
We should fix the codecov token issue - but I'll just push your changes into a different branch to avoid blocking the merge.

**Comment by doc-E-brown** [comment]
Awesome, thanks!

**Comment by hardbyte** [comment]
I've merged a fix to the codecov token bug - if you want to update we *should* see the CI pass.

**Comment by hardbyte** [comment]
Also please note @wilko77's comments in https://github.com/data61/clkhash/pull/343#pullrequestreview-367699607

**Comment by hardbyte** [comment]
> > It's great that you made clkhash work in lambda functions.
> > A typical clkhash user might not know the difference between multi-processing and multi-threading. So this all might be a bit confusing. Do we really need to expose that switch in the CLI? Do you call clkhash in the AWS lambda function via the CLI?
> > I think we should provide more info in the CLI and docs, if we want to keep that switch there.
> 
> We are directly using the generate_clks function in the lambda function not the CLI. I agree about more info within the docs and I will update the PR accordingly. The reason for adding the switch in the CLI was to allow quick testing of the performance drop in using threads as opposed to multiprocesses to see if this is acceptable. I am happy to add this information to the docs as well or just remove the switch altogether.

Cool, lets just remove the CLI switch then good to merge.

File: clkhash/clk.py [hunk 6]
```diff
@@ -75,6 +76,10 @@ def generate_clk_from_csv(input_f,  # type: TextIO
             header but it should not be checked against the schema.
         :param bool progress_bar: Set to `False` to disable the progress
             bar.
+        :param bool use_multiprocessing: Set to `False` to use threading instead
+            of multiprocessing to generate encodings for each record. Set this flag 
+            to `False` if executing on a system that is not capable of spawning
+            subprocesses such as AWS Lambda functions.
```
Comment (hardbyte) [inline clkhash/clk.py]: Picking up the conversation from https://github.com/data61/clkhash/pull/343#discussion_r386796625
@wilko77 @doc-E-brown 
Perhaps a cleaner approach than introducing another rarely used optional variable is to replace `use_multiprocessing` with an executor argument - expect an optional  [`Executor` instance](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor) (with the Process Pool Executor being the default).

Comment (doc-E-brown) [reply clkhash/clk.py]: Sounds like a good approach to me, I'll update soon.

Comment (doc-E-brown) [reply clkhash/clk.py]: Just to touch base, I haven't forgotten about this, just slammed at work.  Will update first chance I get.

Comment (hardbyte) [reply clkhash/clk.py]: Fair enough! The whole world is a bit crazy right now 😷 
Thanks for the update and just let us know if you'd prefer us to implement the suggested changes.

**Pull request closed without merging**